packages/grid/apps/network/src/main/core/node.py

```diff
@@ -72,6 +92,244 @@ def create_network_app(app, args, testing=False):
+                    "value": value,
+                    "sender_address": address,
+                }
+                # url = address + "/association-requests/receive"
+                print("/ws_conn/respond_association_request df", df)
+
+                # response = post(url=url, json=payload)
+                sid = df.loc[address]["sid"]
+
+                ev = event.Event()
+                # TODO: check if message with reply
+                socketio.emit(
+                    "receive_association_request", payload, to=sid,
+                    callback=answer_callback(ev)
+                )
+                timeout = Timeout(10)
+                try:
+                    response = ev.wait()
+                except Timeout:
+                    print("answer timed out")
+                    response = ""
+                finally:
+                    timeout.cancel()
+
+                # TODO: status code from socketio client
+                # response_message = (
+                #     "Association request replied!"
+                #     if response.status_code == 200
+                #     else "Association request could not be replied! Please, try again."
+                # )
+            else:
+                raise AuthorizationError("You're not allowed to create an Association Request!")
+
+            # TODO: error handling
+            response_msg =  RespondAssociationRequestResponse(
+                address=msg.reply_to,
+                status_code=200, # response.status_code,
+                content={"msg": "Association request replied!"}#response_message},
+            )
+            return {"message": _serialize(obj=response_msg, to_bytes=True).decode("ISO-8859-1")}
+
+        # TODO: token required from domain!
+        @app.route("/search-datasets", methods=["GET"])
+        @token_required
+        def search(user: Any):
+            # TODO: send serialized messages instead of creating them here
+            # msg = NetworkSearchMessage(
+            #     address=node.address,
+            #     content={},
+            #     reply_to=node.address
+            # )
+            #
+            # queries = set(msg.content.get("query", []))
+
+            content = request.json
+
+            print("search-datasets", content)
+
+            queries = set(content["queries"])
+            associations = node.association_requests.associations()
+            print("search-datasets associations", associations)
+
+            def filter_domains(url):
+                print("search-datasets filter_domains", df)
+                # datasets = json.loads(requests.get(url + "/data-centric/tensors").text)
+                sid = df.loc[url]["sid"]
+                ev = event.Event()
+                # TODO: check if message with reply
+                socketio.emit(
+                    "get-all-tensors", to=sid,
+                    callback=answer_callback(ev)
+                )
+                timeout = Timeout(10)
```

Comment:
Could you update this to the newest architecture? This always means if waits for a timeout but there s a faster veriosn with 
datasets = None
while datasets is None:
    datasets = ev.wait(timeout=0.1)
Please also adjust everywhere else!

Reply:
Should the max timeout be kept?

e.g. 
```
response = None
timer = 0
while response is None:
    timeout = 0.1
    if (timer + timeout > 10):
        print("Answer timed out")
        break
    else:
        timer += timeout
    response = ev.wait(timeout=timeout)
```

Since this code is highly reused (and long with the addition of timeout handeling) it might be a good idea to create a function "get_response(event, timeout=0.1, max_timeout=10, default_return_value=None)". The question would just be where to add that function since both network and domain could use it various times.

Reply:
Please integrate both, good idea!

Reply:
Is there a place I could implement such a function? Since both Network and Domain use the same implementation it would be prettier if they could share it somehow instead of each having their own implementation.

Reply:
possibly syft/grid/connections/